colorizer_data/converter.py

```diff
@@ -0,0 +1,392 @@
+    backdrop_info: Optional[Dict[str, BackdropMetadata]] = None
+    feature_column_names: Optional[List[str]] = None
+    feature_info: Optional[Dict[str, FeatureInfo]] = None
+    use_json: bool = False
+
+
+def _get_image_from_row(row: pd.DataFrame, config: ConverterConfig) -> AICSImage:
+    zstackpath = row[config["image_column"]]
+    zstackpath = sanitize_path_by_platform(zstackpath)
+    return AICSImage(zstackpath)
+
+
+def _make_frame(
+    frame: pd.DataFrame,
+    scale: float,
+    bounds_arr: Sequence[int],
+    writer: ColorizerDatasetWriter,
+    config: ConverterConfig,
+):
+    start_time = time.time()
+
+    # Get the path to the segmented zstack image frame from the first row (should be the same for
+    # all rows in this group, since they are all on the same frame).
+    row = frame.iloc[0]
+    frame_number = row[config["times_column"]]
+    # Flatten the z-stack to a 2D image.
+    aics_image = _get_image_from_row(row, config)
+    zstack = aics_image.get_image_data("ZYX", S=0, T=0, C=0)
+    seg2d = zstack.max(axis=0)
+
+    # Scale the image and format as integers.
+    seg2d = scale_image(seg2d, scale)
+    seg2d = seg2d.astype(np.uint32)
+
+    # Remap the frame image so the IDs are unique across the whole dataset.
+    seg_remapped, lut = remap_segmented_image(
+        seg2d,
+        frame,
+        config["object_id_column"],
+    )
+
+    writer.write_image(seg_remapped, frame_number)
+    update_bounding_box_data(bounds_arr, seg_remapped)
+
+    time_elapsed = time.time() - start_time
+    logging.info(
+        "Frame {} finished in {:5.2f} seconds.".format(int(frame_number), time_elapsed)
+    )
+
+
+def _make_frames_parallel(
+    grouped_frames: DataFrameGroupBy,
+    scale: float,
+    writer: ColorizerDatasetWriter,
+    config: ConverterConfig,
+):
+    """
+    Generate the images and bounding boxes for each time step in the dataset.
+    """
+    nframes = len(grouped_frames)
+    total_objects = get_total_objects(grouped_frames)
+    logging.info("Making {} frames...".format(nframes))
+
+    with multiprocessing.Manager() as manager:
+        bounds_arr = manager.Array("i", [0] * int(total_objects * 4))
+        with multiprocessing.Pool() as pool:
+            pool.starmap(
+                _make_frame,
+                [
+                    (frame, scale, bounds_arr, writer, config)
+                    for _group_name, frame in grouped_frames
+                ],
+            )
+        writer.write_data(bounds=np.array(bounds_arr, dtype=np.uint32))
+
+
+def _get_data_or_none(
+    dataset: pd.DataFrame, column_name: str
+) -> Union[np.ndarray, None]:
+    if column_name in dataset.columns:
+        return dataset[column_name].to_numpy()
+    else:
+        return None
+
+
+def _write_data(
+    dataset: pd.DataFrame,
+    writer: ColorizerDatasetWriter,
+    config: ConverterConfig,
+):
+    writer.write_data(
+        tracks=_get_data_or_none(dataset, config["track_column"]),
```

Comment:
I feel like this would be cleaner if you could do `config.track_column` etc

Reply:
Looking around this seems to be the standard access pattern for TypedDict?